File: app/src/main/java/net/xzos/UpgradeAll/server/log/LogLiveData.java
package net.xzos.UpgradeAll.server.log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import net.xzos.UpgradeAll.server.ServerContainer;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

class LogLiveData {
    private MutableLiveData<JSONObject> mLogJSONObject = new MutableLiveData<>(new JSONObject());

    void setLogJSONObject(JSONObject logJSONObject) {
        mLogJSONObject.postValue(logJSONObject);
    }

    LiveData<List<String>> getLiveDataLogSortList() {
        return Transformations.map(mLogJSONObject, logJsonObject -> {
            LogUtil logUtil = ServerContainer.AppServer.getLog();
            LogDataProxy logDataProxy = new LogDataProxy(logUtil);
            return logDataProxy.getLogSort();
        });
    }

    LiveData<List<String>> getLiveDataLogObjectIdList(String logSort) {
        return Transformations.map(mLogJSONObject, logJsonObject -> {
            LogUtil logUtil = ServerContainer.AppServer.getLog();
            LogDataProxy logDataProxy = new LogDataProxy(logUtil);
            return logDataProxy.getLogObjectId(logSort);
        });
    }

    LiveData<List<String>> getLiveDataLogList(String[] logObjectTag) {
        return Transformations.map(mLogJSONObject, logJsonObject -> {
            LogUtil logUtil = ServerContainer.AppServer.getLog();
            LogDataProxy logDataProxy = new LogDataProxy(logUtil);
            try {
                return logDataProxy.getLogMessageList(logObjectTag);
            } catch (JSONException e) {
                e.printStackTrace();
                return new ArrayList<>();
            }
        });
    }
}

File: app/src/main/java/net/xzos/upgradeall/server/log/LogLiveData.kt
package net.xzos.upgradeall.server.log

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import net.xzos.dupdatesystem.core.data.json.nongson.ObjectTag
import net.xzos.dupdatesystem.core.log.LogDataProxy
import net.xzos.dupdatesystem.core.log.LogItemData
import net.xzos.upgradeall.utils.notifyObserver


object LogLiveData {

    private var logMap = hashMapOf<ObjectTag, MutableList<LogItemData>>()

    private val mLogMapLiveData = MutableLiveData(logMap)

    fun notifyChange(logMap: HashMap<ObjectTag, MutableList<LogItemData>>) {
        this.logMap = logMap
        mLogMapLiveData.notifyObserver()
    }

    internal val sortList: LiveData<MutableList<String>>
        get() = Transformations.map(mLogMapLiveData) {
            LogDataProxy.logSort
        }

    fun getObjectTagListBySort(logSort: String): LiveData<List<ObjectTag>> {
        return Transformations.map(mLogMapLiveData) {
            LogDataProxy.getObjectTagBySort(logSort)
        }
    }

    fun getLogMassageList(objectTag: ObjectTag): LiveData<List<String>> {
        return Transformations.map(mLogMapLiveData) {
            LogDataProxy.getLogMessageList(objectTag)
        }
    }
}
